### ISSUE TITLE
[Bug]: react/jsx-no-leaked-render reports error on boolean values and tries to convert them

### ISSUE THREAD

**Issue by bytasv** [jsx-eslint/eslint-plugin-react#3908]
### Is there an existing issue for this?

- [x] I have searched the existing issues and my issue is unique
- [x] My issue appears in the command-line and not only in the text editor

### Description Overview

Using following setup:
```
'react/jsx-no-leaked-render': ['error', { validStrategies: ['ternary', 'coerce'] }]
```

![Image](https://github.com/user-attachments/assets/1880de79-5072-4b46-a3e1-4696d6cda266)

The code reports error even though `emptyCollection` is already boolean value. I understand that the plugin does not take into account actual types that TS provides.

Running lint command reports an issue unless I convert the code to this 👇 

![Image](https://github.com/user-attachments/assets/956d666a-56ac-4b14-9820-ed38cd302d82)

Is there a way to prevent this from being "fixed" since it's already a boolean value? 

### eslint-plugin-react version

v7.37.4

### eslint version

v9.18.0

### node version

v22.13.0

**Comment by ljharb** [comment]
When the type information is available, it would be a fine enhancement to use it.

**Comment by shpakkdv** [comment]
I would like to add several simple examples for TS:

`icon` is `object` or `undefined`, and the rule shouldn't report the error in such a case (`undefined` is safe for React to render):

```tsx
interface IconButtonProps {
  icon?: object;
}

const IconButton: FC<IconButtonProps> = ({ icon }) => {
  return (
    <button>
      {icon && <FontAwesomeIcon icon={icon} />}
    </button>
  );
};
```

`isGreetingShown` is always a boolean, and while the rule can't infer the type, it always reports false-positive errors.
Same with `isDeveloper`, but it can be `boolean` or `undefined`, which both are ok for React.

```tsx
interface ComponentProps {
  isGreetingShown: boolean;
  isDeveloper?: boolean;
}

const Component: FC<ComponentProps> = ({ isShown }) => {
  return (
    <div>
      {isGreetingShown && <span>Hello</span>}
      {isDeveloper && <span>, developer!</span>}
    </div>
  );
}
```
